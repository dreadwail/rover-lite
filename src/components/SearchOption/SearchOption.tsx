import React, { PureComponent } from 'react';

import {
  isBooleanOption,
  isChoiceOption,
  isDateOption,
  isIntegerOption,
  isMultipleChoiceOption,
  Option,
} from '../../state/searchOptions/types';
import SearchOptionBoolean from '../SearchOptionBoolean';
import SearchOptionChoice from '../SearchOptionChoice';
import SearchOptionDate from '../SearchOptionDate';
import SearchOptionInteger from '../SearchOptionInteger';
import SearchOptionMultipleChoice from '../SearchOptionMultipleChoice';

export interface SearchOptionProps {
  readonly optionKey: string;
  readonly option: Option;
  readonly value?: any;
  onChange(value: any): void;
}

export default class SearchOption extends PureComponent<SearchOptionProps> {
  public render() {
    const { optionKey, option, value, onChange } = this.props;
    if (isDateOption(option)) {
      return (
        <SearchOptionDate optionKey={optionKey} option={option} value={value} onChange={onChange} />
      );
    }
    if (isIntegerOption(option)) {
      return (
        <SearchOptionInteger
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isBooleanOption(option)) {
      return (
        <SearchOptionBoolean
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isMultipleChoiceOption(option)) {
      return (
        <SearchOptionMultipleChoice
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    if (isChoiceOption(option)) {
      return (
        <SearchOptionChoice
          optionKey={optionKey}
          option={option}
          value={value}
          onChange={onChange}
        />
      );
    }
    return null;
  }
}
