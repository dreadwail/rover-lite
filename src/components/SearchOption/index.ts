export * from './SearchOption';
export { default } from './SearchOption';
