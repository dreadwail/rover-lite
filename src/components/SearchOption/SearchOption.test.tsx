import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

import {
  BasicOption,
  BooleanOption,
  ChoiceOption,
  DateOption,
  IntegerOption,
  MultipleChoiceOption,
} from '../../state/searchOptions/types';

import SearchOption, { SearchOptionProps } from './SearchOption';

describe('SearchOption', () => {
  let props: SearchOptionProps;
  let rendered: ShallowWrapper;

  beforeEach(() => {
    props = {
      optionKey: 'test-option-key',
      onChange: jest.fn().mockName('onChange'),
      option: {
        type: 'unknown',
        required: false,
        label: 'unknown-label',
        category: {
          categoryId: 'unknown-category-id',
          rank: 0,
          displayName: 'Unknown Category Name',
        },
        rank: 0,
      },
    };
  });

  describe('with a date option', () => {
    beforeEach(() => {
      const option: DateOption = {
        type: 'date',
        required: false,
        label: 'test-date-label',
        category: {
          categoryId: 'test-date-category-id',
          rank: 0,
          displayName: 'Test Date Category',
        },
        rank: 0,
      };
      const value = new Date(1970, 0, 1);
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with an integer option', () => {
    beforeEach(() => {
      const option: IntegerOption = {
        type: 'integer',
        minValue: 0,
        maxValue: 50,
        required: false,
        label: 'test-integer-label',
        category: {
          categoryId: 'test-integer-category-id',
          rank: 0,
          displayName: 'Test Integer Category',
        },
        rank: 0,
      };
      const value = 42;
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with a boolean option', () => {
    beforeEach(() => {
      const option: BooleanOption = {
        type: 'boolean',
        required: false,
        label: 'test-integer-label',
        category: {
          categoryId: 'test-integer-category-id',
          rank: 0,
          displayName: 'Test Integer Category',
        },
        rank: 0,
      };
      const value = true;
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with a multiple choice option', () => {
    beforeEach(() => {
      const option: MultipleChoiceOption = {
        type: 'multiple choice',
        choices: [
          { value: 'test-choice-1', displayName: 'Test Choice 1' },
          { value: 'test-choice-2', displayName: 'Test Choice 2' },
        ],
        required: false,
        label: 'test-multiple-choice-label',
        category: {
          categoryId: 'test-multiple-choice-category-id',
          rank: 0,
          displayName: 'Test Multiple Choice Category',
        },
        rank: 0,
      };
      const value = ['test-choice-1', 'test-choice-2'];
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with a choice option', () => {
    beforeEach(() => {
      const option: ChoiceOption = {
        type: 'choice',
        choices: [
          { value: 'test-choice-1', displayName: 'Test Choice 1' },
          { value: 'test-choice-2', displayName: 'Test Choice 2' },
        ],
        required: false,
        label: 'test-choice-label',
        category: {
          categoryId: 'test-choice-category-id',
          rank: 0,
          displayName: 'Test Choice Category',
        },
        rank: 0,
      };
      const value = 'test-choice-1';
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders correctly', () => {
      expect(rendered).toMatchSnapshot();
    });
  });

  describe('with an unhandled option', () => {
    beforeEach(() => {
      const option: BasicOption = {
        type: 'nope',
        required: false,
        label: 'test-unknown-label',
        category: {
          categoryId: 'test-unknown-category-id',
          rank: 0,
          displayName: 'Test Unknown Category',
        },
        rank: 0,
      };
      const value = {};
      props = { ...props, option, value };
      rendered = shallow(<SearchOption {...props} />);
    });

    it('renders nothing', () => {
      expect(rendered.isEmptyRender()).toBe(true);
    });
  });
});
