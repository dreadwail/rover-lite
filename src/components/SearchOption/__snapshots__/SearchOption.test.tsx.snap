// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchOption with a boolean option renders correctly 1`] = `
<SearchOptionBoolean
  onChange={[MockFunction onChange]}
  option={
    Object {
      "category": Object {
        "categoryId": "test-integer-category-id",
        "displayName": "Test Integer Category",
        "rank": 0,
      },
      "label": "test-integer-label",
      "rank": 0,
      "required": false,
      "type": "boolean",
    }
  }
  optionKey="test-option-key"
  value={true}
/>
`;

exports[`SearchOption with a choice option renders correctly 1`] = `
<SearchOptionChoice
  onChange={[MockFunction onChange]}
  option={
    Object {
      "category": Object {
        "categoryId": "test-choice-category-id",
        "displayName": "Test Choice Category",
        "rank": 0,
      },
      "choices": Array [
        Object {
          "displayName": "Test Choice 1",
          "value": "test-choice-1",
        },
        Object {
          "displayName": "Test Choice 2",
          "value": "test-choice-2",
        },
      ],
      "label": "test-choice-label",
      "rank": 0,
      "required": false,
      "type": "choice",
    }
  }
  optionKey="test-option-key"
  value="test-choice-1"
/>
`;

exports[`SearchOption with a date option renders correctly 1`] = `
<SearchOptionDate
  onChange={[MockFunction onChange]}
  option={
    Object {
      "category": Object {
        "categoryId": "test-date-category-id",
        "displayName": "Test Date Category",
        "rank": 0,
      },
      "label": "test-date-label",
      "rank": 0,
      "required": false,
      "type": "date",
    }
  }
  optionKey="test-option-key"
  value={1970-01-01T08:00:00.000Z}
/>
`;

exports[`SearchOption with a multiple choice option renders correctly 1`] = `
<SearchOptionMultipleChoice
  onChange={[MockFunction onChange]}
  option={
    Object {
      "category": Object {
        "categoryId": "test-multiple-choice-category-id",
        "displayName": "Test Multiple Choice Category",
        "rank": 0,
      },
      "choices": Array [
        Object {
          "displayName": "Test Choice 1",
          "value": "test-choice-1",
        },
        Object {
          "displayName": "Test Choice 2",
          "value": "test-choice-2",
        },
      ],
      "label": "test-multiple-choice-label",
      "rank": 0,
      "required": false,
      "type": "multiple choice",
    }
  }
  optionKey="test-option-key"
  value={
    Array [
      "test-choice-1",
      "test-choice-2",
    ]
  }
/>
`;

exports[`SearchOption with an integer option renders correctly 1`] = `
<SearchOptionInteger
  onChange={[MockFunction onChange]}
  option={
    Object {
      "category": Object {
        "categoryId": "test-integer-category-id",
        "displayName": "Test Integer Category",
        "rank": 0,
      },
      "label": "test-integer-label",
      "maxValue": 50,
      "minValue": 0,
      "rank": 0,
      "required": false,
      "type": "integer",
    }
  }
  optionKey="test-option-key"
  value={42}
/>
`;
